### ISSUE TITLE
Return value of higher order functions looses excess property checks

### ISSUE THREAD

**Issue by essenmitsosse** [microsoft/TypeScript#39635]
**TypeScript Version:**  3.9.2 (doesn't work in any version I tested, including 4)

**Search Terms:**
higher order function return value object excess property check

**Code**

```ts
interface Foo {
	foo: {
		bar: number
	}
}

// Error for unknown property buz
const getFoo = (): Foo  => ({
	foo: {
		bar: 1,
		buz: 2,
	}
})

// Typeing should be the same. Autocomplete works, but no error is shown for the excessive property buz
const getFoo3 = (): () => Foo  => () => ({
	foo: {
		bar: 1,
		buz: 2
	}
})

// Same problem as above. The issue doesn't seem to be directly related to higher order functions
const getFoo2: () => Foo = ()  => ({
	foo: {
		bar: 1,
		buz: 2,
	}
})
```

**Expected behavior:**
Error should be thrown for the excess property in all cases.

**Actual behavior:**
Error for excess properties is only thrown, when the return value of the function is directly typed.

**[Playground Link](https://www.typescriptlang.org/play/?ssl=27&ssc=10&pln=26&pc=10#code/JYOwLgpgTgZghgYwgAgGIHt3IN4CgCQMmAXDgfgEZxSkgCuAthdAQL67u4D0XyAolCjooyIiLogA1iHQB3EMgAOQxdDABPZBToAvXAnQgAzmGQBzCGAxYAvMgAUASlLXkyGwD4HeQiTL5KalIARgAacm0dUgAmcPx2Vkdcbl4AFXVVUDNkIwALdDoAGwATLRQwXJQjOAYIADpkAEE6MHQDBkVCyxRZYUkjUK0W5BlkaCERYCMc-PlRYWQKlAgADyQjI2AANxRldFUoDSG9A2NTCytMAGZ3B2c79y9XR4fPbwIidFIfQJpkMIiuhibA4SRSyAAyjVdkIKF0GMg4NM4BR0DsGqlKsgpkY6ChiugIEYQAByUxGCAQBGtMrIYrAKAQBBgQqaRmFOCQUo03LAMyVETCYrQUQSZnAQxGfSS86WazRUhOF6uOxKl72Hyfb4RIL-OKUIHIWIgxK4IA)**

**Related Issues:** -

**Comment by RyanCavanaugh** [comment]
Duplicate #241
